presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/TestMemo.java

```diff
@@ -0,0 +1,265 @@
+    /*
+       From: X -> Z
+       To:   X -> Y -> Z
+     */
+    @Test
+    public void testInsertNode()
+            throws Exception
+    {
+        PlanNode z = node();
+        PlanNode x = node(z);
+
+        Memo memo = new Memo(idAllocator, x);
+
+        assertEquals(memo.getGroupCount(), 2);
+
+        int zGroup = getChildGroup(memo, memo.getRootGroup());
+        PlanNode y = node(memo.getNode(zGroup));
+        memo.replace(zGroup, y, "rule");
+
+        assertEquals(memo.getGroupCount(), 3);
+
+        assertMatchesStructure(
+                memo.extract(),
+                node(x.getId(),
+                        node(y.getId(),
+                                node(z.getId()))));
+    }
+
+    /*
+      From: X -> Y -> Z
+      To:   X --> Y1' --> Z
+              \-> Y2' -/
+     */
+    @Test
+    public void testMultipleReferences()
+            throws Exception
+    {
+        PlanNode z = node();
+        PlanNode y = node(z);
+        PlanNode x = node(y);
+
+        Memo memo = new Memo(idAllocator, x);
+        assertEquals(memo.getGroupCount(), 3);
+
+        int yGroup = getChildGroup(memo, memo.getRootGroup());
+
+        PlanNode rewrittenZ = memo.getNode(yGroup).getSources().get(0);
+        PlanNode y1 = node(rewrittenZ);
+        PlanNode y2 = node(rewrittenZ);
+
+        PlanNode newX = node(y1, y2);
+        memo.replace(memo.getRootGroup(), newX, "rule");
+        assertEquals(memo.getGroupCount(), 4);
+
+        assertMatchesStructure(
+                memo.extract(),
+                node(newX.getId(),
+                        node(y1.getId(), node(z.getId())),
+                        node(y2.getId(), node(z.getId()))));
+    }
+
+    private static void assertMatchesStructure(PlanNode actual, PlanNode expected)
+    {
+        assertEquals(actual.getClass(), expected.getClass());
+        assertEquals(actual.getId(), expected.getId());
+        assertEquals(actual.getSources().size(), expected.getSources().size());
+
+        for (int i = 0; i < actual.getSources().size(); i++) {
+            assertMatchesStructure(actual.getSources().get(i), expected.getSources().get(i));
+        }
+    }
+
+    private int getChildGroup(Memo memo, int group)
+    {
+        PlanNode node = memo.getNode(group);
+        GroupReference child = (GroupReference) node.getSources().get(0);
+
+        return child.getGroupId();
+    }
+
+    private GenericNode node(PlanNodeId id, PlanNode left, PlanNode right)
+    {
+        return new GenericNode(id, ImmutableList.of(left, right));
+    }
+
+    private GenericNode node(PlanNode left, PlanNode right)
+    {
+        return node(idAllocator.getNextId(), left, right);
+    }
+
+    private GenericNode node(PlanNodeId id, PlanNode source)
+    {
+        return new GenericNode(id, ImmutableList.of(source));
+    }
+
+    private GenericNode node(PlanNode source)
+    {
+        return node(idAllocator.getNextId(), source);
+    }
+
+    private GenericNode node(PlanNodeId id)
```

Comment:
Ah, duh! I renamed the old filter/project/join/values methods and forgot to deal with the duplication :)